builder/builder.go

```diff
@@ -379,3 +395,82 @@ func parseImageNameFromArgs(cmdArgs string) string {
 	}
 	return cmdArgs[:idx]
 }
+
+func resolveCustomRegistryCredentials(ctx context.Context, credentials []*graph.RegistryCredential) ([]*ResolvedRegistryCred, error) {
+	resolvedCreds := make(map[string]string)
+	var unresolvedCreds []*graph.Secret
+
+	for _, cred := range credentials {
+		usernameKey := cred.Registry + "_username"
+		passwordKey := cred.Registry + "_password"
+		resolvedCreds[usernameKey] = ""
+		resolvedCreds[passwordKey] = ""
+		isMSI := false
+
+		switch cred.UsernameType {
+		case graph.Opaque:
+			resolvedCreds[usernameKey] = cred.Username
+		case graph.VaultSecret:
+			unresolvedCreds = append(unresolvedCreds, &graph.Secret{
+				InternalID:  usernameKey,
+				Akv:         cred.Username,
+				MsiClientID: cred.Identity,
+			})
+		case "":
+			isMSI = true
+		}
+
+		switch cred.PasswordType {
+		case graph.Opaque:
+			resolvedCreds[passwordKey] = cred.Password
+		case graph.VaultSecret:
+			unresolvedCreds = append(unresolvedCreds, &graph.Secret{
+				InternalID:  passwordKey,
+				Akv:         cred.Password,
+				MsiClientID: cred.Identity,
+			})
+		}
+
+		if isMSI {
+			resolvedCreds[usernameKey] = "00000000-0000-0000-0000-000000000000"
+			unresolvedCreds = append(unresolvedCreds, &graph.Secret{
+				InternalID:    passwordKey,
+				MsiClientID:   cred.Identity,
+				ArmResourceID: cred.ArmResourceID,
+			})
+		}
+	}
+
+	secretResolver, err := templating.NewSecretResolver(nil, templating.DefaultSecretResolveTimeout)
```

Comment:
You may consider to refactor SecretResolver to a separate module. It's strange to reference anything from templating in this function.